### PR TITLE
[Feature] Recurring expenses

### PR DESCRIPTION
**Pull request by Civolilah**
Here are screenshots of implemented feature [Recurring expenses]:

Screenshot 2022-12-20 at 15 08 41

Screenshot 2022-12-20 at 14 54 54

Screenshot 2022-12-20 at 14 56 42

@turbo124 Just a few things about translation files. We are missing two keywords associated with this feature: `started_recurring_expense` and `stopped_recurring_expense`. If you look at the third screenshot, `clone_to_expense` which gives us the result 'Clone to expense', I suggest we update it to 'Clone to Expense' to be consistent with the others. Also, for the number label on the Edit page I used `expense_number` instead of `recurring_expense_number`. IMO it was enough, but let me know what you think.

### CODE REVIEW

**Comment by Civolilah** [comment]
> 

@turbo124 I changed the translation keywords for the cases you mentioned. But, IMO, maybe just using Start/Stop is a bit unusual. Rather to use Started/Stopped? In that case, we have already `stopped` in the translation files and just need to add the `started`. When we talk about recurring invoices, this is fully implemented, `started_recurring_invoice` with message 'Succesfully started recurring invoice' for example. But I agree with you that maybe just using Started/Stopped would be enough.

**Comment by turbo124** [comment]
We use start/stop for this in recurring invoices. I think this is the correct word for this setting.

**Comment by Civolilah** [comment]
> We use start/stop for this in recurring invoices. I think this is the correct word for this setting.

@turbo124 Okay, the toaster message has been updated!

**Comment by turbo124** [comment]
@Civolilah for recurring expenses, the only required field is frequency Id, this should automatically default to monthly.

**Comment by Civolilah** [comment]
> @Civolilah for recurring expenses, the only required field is frequency Id, this should automatically default to monthly.

@turbo124 Sure I missed it, now it's adjusted. Just to let you know, I double checked the recurring invoices and saw that blank query returns `frequency_id=5` (Monthly), in the case of recurring expenses we don't have that parameter set to any value. I set this default value manually, but if you think it's better to return through an empty query, I can simply remove this manullay setting.

**Comment by turbo124** [comment]
@Civolilah 

I"ve just pulled this in, i'm still unable to save a blank recurring expense, the save button is blanked out.

**Comment by Civolilah** [comment]
> 

@turbo124 The button is disabled when no client is selected. I saw the same kind of disablement on another feature, so I just replicated it. You will now be able to create blank recurring expense, I have removed this kind of validation.

**Comment by turbo124** [comment]
@Civolilah looks good.

One small change, endless recurring expenses have a frequency of -1, can we please alias this to the 'endless' translation string please.

**Comment by turbo124** [comment]
@Civolilah 

One thing I have just notifced, the recurring expenses do not have the column functionality that the other entities have. Please ensure the columns functionality is also implemented in this PR please.

**Comment by Civolilah** [comment]
> 

done

File: src/pages/recurring-expenses/common/hooks.tsx [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+  'vendor',
+  'client',
+  'date',
+  'amount',
+  'public_notes',
+  'entity_state',
+  'archived_at',
+  //   'assigned_to', @Todo: Need to resolve relationship
+  //   'category', @Todo: Need to resolve relationship
+  'created_at',
```
Comment (Civolilah) [inline src/pages/recurring-expenses/common/hooks.tsx:58]: I kept this comments same as for expense entity, please let me know if you have any updates here.

**Comment by turbo124** [comment]
@Civolilah @beganovich LGTM

File: src/pages/expenses/common/hooks.tsx [hunk 2]
```diff
@@ -65,6 +74,47 @@ export const expenseColumns = [
...
 type ExpenseColumns = typeof expenseColumns[number];
 
+export function useActions() {
+  const [t] = useTranslation();
+
+  const navigate = useNavigate();
+
+  const [, setExpense] = useAtom(expenseAtom);
+
+  const [, setRecurringExpense] = useAtom(recurringExpenseAtom);
```
Comment (beganovich) [inline src/pages/expenses/common/hooks.tsx:84]: To skip the getter for the state, you can use `useUpdateAtom`.

```ts
const setExpense = useUpdateAtom(expenseAtom);
```

https://jotai.org/docs/utils/use-update-atom

File: src/pages/expenses/create/Create.tsx [hunk 3]
```diff
@@ -26,42 +26,53 @@ import { TaxSettings } from './components/Taxes';
...
-  ) => {
-    setExpense((expense) => expense && { ...expense, [property]: value });
-  };
+  useEffect(() => {
+    if (mounted) {
+      return () => setExpense(undefined);
+    } else {
+      mounted = true;
+    }
+  }, []);
```
Comment (beganovich) [inline src/pages/expenses/create/Create.tsx:71]: Why?

File: src/pages/expenses/create/Create.tsx [hunk 4]
```diff
@@ -26,42 +26,53 @@ import { TaxSettings } from './components/Taxes';
...
 export function Create() {
   const [t] = useTranslation();
 
   const navigate = useNavigate();
 
   const { documentTitle } = useTitle('new_expense');
+
   const { data } = useBlankExpenseQuery();
 
+  let mounted = false;
```
Comment (beganovich) [inline src/pages/expenses/create/Create.tsx:42]: Generally you should avoid this style of variables, as they aren't reactive (won't re-render).

File: src/pages/recurring-expenses/create/Create.tsx [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    }
+  }, [data]);
+
+  useEffect(() => {
+    if (mounted) {
+      return () => setRecurringExpense(undefined);
+    } else {
+      mounted = true;
+    }
+  }, []);
```
Comment (beganovich) [inline src/pages/recurring-expenses/create/Create.tsx:71]: Please let me know why we do this.

**Review by beganovich — CHANGES_REQUESTED** [review]
Looks good, small changes are needed.

The documents tab is taking full page width (same with expenses). Can we please do 2/3 of the page?

image

File: src/pages/expenses/common/hooks.tsx (hunk 2 above)
Comment (Civolilah) [reply src/pages/expenses/common/hooks.tsx:84]: replaced

File: src/pages/expenses/create/Create.tsx (hunk 4 above)
Comment (Civolilah) [reply src/pages/expenses/create/Create.tsx:42]: As we discussed on the Slack this part is removed from technical reasons

File: src/pages/expenses/create/Create.tsx (hunk 3 above)
Comment (Civolilah) [reply src/pages/expenses/create/Create.tsx:71]: Same as above, removed

File: src/pages/recurring-expenses/create/Create.tsx (hunk 5 above)
Comment (Civolilah) [reply src/pages/recurring-expenses/create/Create.tsx:71]: removed

**Comment by Civolilah** [comment]
> 

@beganovich Including other small changes in the code the UI for document tab is also slightly changed and it looks like this:

Screenshot 2023-01-02 at 16 25 54

**Comment by beganovich** [comment]
Looks good, thanks 👍

**Pull request merged**